core/src/main/java/io/parsingdata/metal/data/ImmutableList.java

```diff
@@ -59,9 +63,13 @@ private ImmutableList(final T head, final ImmutableList<T> tail) {
 
     public ImmutableList<T> add(final ImmutableList<T> list) {
         checkNotNull(list, "list");
-        if (list.isEmpty()) { return this; }
```

Comment:
Oh good point. The duplication can prevent the creation of an unnecessary trampoline :) In this case I will go for maintainability (= prevent duplication).